src/jdk.jshell/share/classes/jdk/internal/jshell/tool/ConsoleIOContext.java

```diff
@@ -977,7 +977,20 @@ public void perform(LineReaderImpl in) throws IOException {
     public synchronized int readUserInput() throws IOException {
         if (pendingBytes == null || pendingBytes.length <= pendingBytesPointer) {
             char userChar = readUserInputChar();
-            pendingBytes = String.valueOf(userChar).getBytes();
+            StringBuilder dataToConvert = new StringBuilder();
```

Comment:
FWIW I think we can avoid using StringBuilder:

```java
char[] dataToConvert = { useChar, '\0' };
// if (...) {
// ...
// if (...) {
// ...
dataToConvert[1] = lowSurrogate;
// }
// ...
// }
// low-surrogate code unit never be null char
pendingBytes = dataToConvert[1] != '\0' ? String.valueOf(dataToConvert) : String.valueOf(dataToConvert[0]);
```

Reply:
We could do something like that, although I wrote it as I wrote it mostly because that's more clearly correct. Although the current tests probably cover all the cases, so with a bit of work, we probably could eliminate the (explicit) array completely.

Overall, on most places, it is usually not necessary to be too clever - the VM can optimize and eliminate allocations if needed.

I'll leave it up to Adam and/or Christian whether they would prefer a slightly more complex code with less (explicit/visible) allocation.

Reply:
You can integrate the current code with StringBuilder.